### DISCUSSION TITLE
Using MJX to accelerate humanoid standing-balance planning with GPU

### DISCUSSION THREAD

**Discussion by RensDur** [google-deepmind/mujoco#2285] [Asking for Help] [answered]
### Intro

Dear mujoco enthousiasts,

I'm a MSc-student Computer Science at the Delft University of Technology. Although I understand a lot of what's happening inside of mujoco, I don't have a ton of experience with it and I have an interesting problem to solve.

### My setup

Currently, I'm running MuJoCo 3.1.4 in C++ on Ubuntu 22.04 and I'm working on a humanoid model. The goal is to let the humanoid perform a self-balanced stand through a customized residual that's based on, among other things, Center of Mass (CoM), joint velocity and amount of joint control input.

The balanced stand runs entirely on the CPU and my goal is to move the planning part of this simulation to the GPU using MJX. Aside from alleviating the load on the CPU while running multiple planning simulations simultaneously, I'm hoping to be able to run (a lot) more of these simultaneous sampling simulations on the GPU to improve the planner's balancing performance.

### My question

My question boils down to this: How do I use MJX in a C++ MuJoCo environment? On [this page](https://mujoco.readthedocs.io/en/stable/mjx.html), I could only find installation and coding instructions for python. How do I install and build MJX and include it in this C++ project?

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by amacati** [comment, marked as answer]
As the docs explain, MJX is a reimplementation of MuJoCo's simulation pipeline in Python using JAX, so you cannot use it from C++. Your best option is probably to port your environment to Python with the official bindings.

**Reply by yuvaltassa** [reply]
Not clear that you want MJX at all for this, BTW. MJX has high throughput because you can use very wide batches, but a single step on a single env is much faster on CPU. So if I understood your use case you should keep using CPU.
